File: portfolio/projects/D_INSA/INSA-salary-usb/backend/app/schemas/salary_audit.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, model_validator

SalaryAction = Literal[
    "OPEN",
    "UNLOCK_FAIL",
    "SAVE",
    "CREATE_VAULT",
    "LOCK",
    "PASSWORD_CHANGE",
    "DELEGATE",
    "IP_DENIED",
    "POLICY_CHANGE",
]


class AccessLogCreate(BaseModel):
    """클라이언트가 보낼 수 있는 값은 이 셋뿐이다.

    user_id·ip_address·user_agent·occurred_at은 서버가 채운다.
    """

    action: SalaryAction
    record_count: Optional[int] = None
    target_user_id: Optional[int] = None

    @model_validator(mode="after")
    def target_only_for_delegate(self) -> "AccessLogCreate":
        if self.target_user_id is not None and self.action != "DELEGATE":
            raise ValueError("target_user_id는 DELEGATE에만 사용할 수 있습니다")
        return self


class AccessLogRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    action: str
    target_user_id: Optional[int]
    record_count: Optional[int]
    ip_address: Optional[str]
    user_agent: Optional[str]
    occurred_at: datetime


class AccessLogListResponse(BaseModel):
    items: list[AccessLogRow]
    total: int

File: portfolio/projects/D_INSA/INSA-salary-usb/frontend/src/api/salaryAudit.ts
import { useQuery } from "@tanstack/react-query";
import apiClient from "./client";

export type SalaryAction =
  | "OPEN"
  | "UNLOCK_FAIL"
  | "SAVE"
  | "CREATE_VAULT"
  | "LOCK"
  | "PASSWORD_CHANGE"
  | "DELEGATE"
  | "IP_DENIED"
  | "POLICY_CHANGE";

export interface AccessLogRow {
  id: number;
  user_id: number;
  action: SalaryAction;
  target_user_id: number | null;
  record_count: number | null;
  ip_address: string | null;
  user_agent: string | null;
  occurred_at: string;
}

/**
 * 접근 기록을 남긴다. 1회 재시도하고, 실패해도 절대 throw하지 않는다.
 *
 * 파일은 INSA 밖에서도 열 수 있으므로 로그가 완전성을 보증하지 못한다.
 * 로그 서버 장애로 급여 업무가 멈추는 쪽이 더 나쁘다 (스펙 §10.4).
 */
export async function recordAccess(
  action: SalaryAction,
  opts?: { recordCount?: number; targetUserId?: number },
): Promise<void> {
  const body = {
    action,
    record_count: opts?.recordCount ?? null,
    target_user_id: opts?.targetUserId ?? null,
  };
  for (let attempt = 0; attempt < 2; attempt += 1) {
    try {
      await apiClient.post("/salary/access-logs", body);
      return;
    } catch (error) {
      if (attempt === 1) {
        console.error("연봉 접근 기록 전송에 실패했습니다:", error);
      }
    }
  }
}

export interface AccessLogFilters {
  userId?: number;
  action?: SalaryAction;
}

export function useAccessLogs(filters: AccessLogFilters) {
  return useQuery({
    queryKey: ["salary", "access-logs", filters],
    queryFn: async () => {
      const { data } = await apiClient.get<{
        items: AccessLogRow[];
        total: number;
      }>("/salary/access-logs", {
        params: { user_id: filters.userId, action: filters.action },
      });
      return data;
    },
  });
}
